sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -110,6 +110,31 @@ trait CheckAnalysis {
               s"filter expression '${f.condition.sql}' " +
                 s"of type ${f.condition.dataType.simpleString} is not a boolean.")
 
+          case f @ Filter(condition, child) =>
+            // Make sure no correlated predicate is in an OUTER join, because this could change the
+            // semantics of the join.
+            lazy val attributes: Set[Expression] = child.output.toSet
+            def checkCorrelatedPredicates(p: PredicateSubquery): Unit = p.query.foreach {
+              case j @ Join(left, right, jt, _) if jt != Inner =>
+                j.transformAllExpressions {
+                  case e if attributes.contains(e) =>
+                    failAnalysis(s"Accessing outer query column is not allowed in outer joins: $e")
+                }
+              case _ =>
+            }
+            splitConjunctivePredicates(condition).foreach {
+              case p: PredicateSubquery =>
+                checkCorrelatedPredicates(p)
+              case Not(InSubQuery(_, query)) if query.output.exists(_.nullable) =>
+                failAnalysis("NOT IN with nullable subquery is not supported. " +
```

Comment:
Yeah, this is possible. It would require us to move the entire rule to the SparkStrategy (it will be messy to pattern match a rewritten NAAJ). The downside of this is, is that SparkStrategy can throw  Analysis error which should have been thrown before.